nominal/core/stream_v2.py

```diff
@@ -0,0 +1,219 @@
+from dataclasses import dataclass, field
+from datetime import datetime, timedelta
+from types import TracebackType
+from typing import Callable, Sequence, Type
+
+from typing_extensions import Self
+
+from nominal.core._clientsbunch import ProtoWriteService
+from nominal.core.queueing import (
+    BackpressureQueue,
+    BlockingQueue,
+    QueueShutdown,
+    ReadQueue,
+    iter_queue,
+    spawn_batching_thread,
+)
+from nominal.core.stream import BatchItem
+
+# if TYPE_CHECKING:
+from nominal.core.worker_pool import WorkerContext, worker_init
+from nominal.ts import IntegralNanosecondsUTC
+
+logger = logging.getLogger(__name__)
+
+
+class BackpressureMode(enum.Enum):
+    BLOCK = "block"
+    DROP_NEWEST = "drop_newest"
+    DROP_OLDEST = "drop_oldest"
+
+
+@dataclass()
+class WriteStreamV2:
+    _process_batch: Callable[[Sequence[BatchItem]], None]
+    max_batch_size: int
+    max_wait: timedelta
+    max_queue_size: int
+    backpressure_mode: BackpressureMode
+    _item_queue: BackpressureQueue[BatchItem | QueueShutdown] = field(init=False)
+    _batch_queue: ReadQueue[Sequence[BatchItem]] | None = field(default=None)
+    _batch_thread: threading.Thread | None = field(default=None)
+    _process_thread: threading.Thread | None = field(default=None)
+    _executor: concurrent.futures.Executor | None = field(default=None)
+
+    @classmethod
+    def create(
+        cls,
+        nominal_data_source_rid: str,
+        process_batch: Callable[[Sequence[BatchItem]], None],
+        max_batch_size: int,
+        max_wait: timedelta,
+        max_queue_size: int,
+        backpressure_mode: BackpressureMode,
+        client_factory: Callable[[], ProtoWriteService],
+        auth_header: str,
+        max_workers: int,
+    ) -> Self:
+        """Create a new WriteStreamV2 instance.
+
+        Args:
+            process_batch: Function to process batches of items
+            max_batch_size: How many items to accumulate before processing
+            max_wait: Maximum time to wait before processing a partial batch
+            max_queue_size: Maximum number of items that can be queued (0 for unlimited)
+            backpressure_mode: How to handle queue overflow:
+                - BLOCK: Block until space is available (default)
+                - DROP_NEWEST: Drop new items when queue is full
+                - DROP_OLDEST: Drop oldest items when queue is full (ring buffer)
+            executor: executor for parallel batch processing.
+            client_factory: Factory function to create ProtoWriteService instances
+            nominal_data_source_rid: Nominal data source rid
+            auth_header: Authentication header
+            max_workers: Maximum number of worker threads for parallel processing
+        """
+        context = WorkerContext(
+            nominal_data_source_rid=nominal_data_source_rid,
+            auth_header=auth_header,
+            proto_write=client_factory(),
```

Comment:
this is safe? I'd imagine you should probs pass just the base params here and construct the client yourself on the other end, I'm not sure if this would have any bad side-effects as the requests client is stateful